processor/groupbytraceprocessor/factory.go

```diff
@@ -0,0 +1,107 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package groupbytraceprocessor
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configerror"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/consumer"
+)
+
+const (
+	// typeStr is the value of "type" for this processor in the configuration.
+	typeStr configmodels.Type = "groupbytrace"
+)
+
+var (
+	defaultWaitDuration   = time.Second
+	defaultNumTraces      = 1_000_000
+	defaultDiscardOrphans = false
+	defaultStoreInDisk    = false
+
+	errDiskStorageNotSupported    = fmt.Errorf("option 'disk storage' not supported in this release")
+	errDiscardOrphansNotSupported = fmt.Errorf("option 'discard orphans' not supported in this release")
+)
+
+// Factory is the factory for the processor.
+type Factory struct {
+}
+
+var _ component.ProcessorFactory = (*Factory)(nil)
+
+// Type gets the type of the config created by this factory.
+func (f *Factory) Type() configmodels.Type {
+	return typeStr
+}
+
+// CreateDefaultConfig creates the default configuration for the processor.
+func (f *Factory) CreateDefaultConfig() configmodels.Processor {
+	return &Config{
+		ProcessorSettings: configmodels.ProcessorSettings{
+			TypeVal: typeStr,
+			NameVal: string(typeStr),
+		},
+		NumTraces:    defaultNumTraces,
+		WaitDuration: defaultWaitDuration,
+
+		// not supported for now
+		DiscardOrphans: defaultDiscardOrphans,
+		StoreInDisk:    defaultStoreInDisk,
+	}
+}
+
+// CreateTraceProcessor creates a trace processor based on this config.
+func (f *Factory) CreateTraceProcessor(
+	_ context.Context,
+	params component.ProcessorCreateParams,
+	nextConsumer consumer.TraceConsumer,
+	cfg configmodels.Processor) (component.TraceProcessor, error) {
+
+	oCfg := cfg.(*Config)
+
+	var st storage
+	if oCfg.StoreInDisk {
+		return nil, errDiskStorageNotSupported
+	}
+	if oCfg.DiscardOrphans {
+		return nil, errDiscardOrphansNotSupported
+	}
+
+	logger := params.Logger
+	if logger == nil {
+		logger = zap.NewNop()
```

Comment:
Setting logger is responsibility of the code invoking `CreateTraceProcessor`.
Currently, logger is never nil: https://github.com/open-telemetry/opentelemetry-collector/blob/db926fd7f488bfabe0a4558578b2d59c48cc56a8/service/builder/pipelines_builder.go#L159

If you need this for unit tests, just init `ProcessorCreateParams` with nop logger in the test.

Reply:
"Currently" being the keyword. I guess I'm just too used to do defensive programming, but I can surely remove it if necessary.

Reply:
None of the existing processors check for that, so "currently" is an implicit contract anyway that would break all components if changed to nil, so I would just remove it.
Go doesn't support null safety, so I think it's fine to just panic with nil pointer dereference wherever the logger is actually used.

Reply:
Removed.